### DISCUSSION TITLE
Computed Properties vs derive

### DISCUSSION THREAD

**Discussion by alexander-entin** [pmndrs/valtio#731] [General]
**Computed Property**
1. Only works as expected when depending on the same object's properties only (see below code)
2. Synchronous
3. Is computed on each use via proxy
4. Is computed upon snapshotting only if the containing object has changed

**`derive`**
1. Can depend on multiple proxies
2. Asynchronous by default
3. Is computed on a dependency change

Is this correct?

If so, then most devs will start with Computed Properties, learn #1 the hard way and switch to `derive`.

Will it be better to deemphasize Computed Properties and promote `derive`? Or at least document the differences clearly.

````javascript
p = proxy({
    a: { x: 1 },
    b: {
        get y() {
            console.log('computing b.y', p.a.x)
            return p.a.x
        }
    },
})
s = snapshot(p)
// -> computing b.y 1
p.a.x = 2
console.log('proxy', { ...p.a }, { ...p.b })
// -> computing b.y 2
// -> proxy {x: 2} {y: 2}
s = snapshot(p)
console.log('snapshot', s.a, s.b)
// -> snapshot {x: 2} {y: 1}  <--------------- OOOPS
````

**Comment by dai-shi** [comment]
I think if we use computed properties, we should use `this`. How does that change your understanding and example?

> Will it be better to deemphasize Computed Properties and promote derive? Or at least document the differences clearly.

I don't disagree.

**Reply by alexander-entin** [reply]
> I think if we use computed properties, we should use `this`. How does that change your understanding and example?

My example can't be rewritten using `this`(?).

1. Computed properties are snapshotted
2. After a change, only changed branches are re-snapshotted
3. So, a computed property in an unchanged branch, which depends on a changed branch, will be stale in the new snapshot

The choice is:
1. Leave the implementation as is but clearly state this caveat in the docs
2. Make it work as expected 

Maybe track references when computing computed props? Looks complex. Is it?

Maybe re-compute _all_ computed props while snapshotting, even on unchanged branches? Looks costly.

Maybe always re-compute computed props on snapshots (as it is done for proxies)? But computed props referencing proxy will now be fuzzy. And the recommendation should be to always use `this`. Which makes a simple use case a bit worse. And doesn't enable a more complex case of referencing other branches, as it can't be done via `this`.

Now, I think I understand your recommendation on `this`:
- Use `this`
- Otherwise use `derive`

Still, will it be better to recommend:
- Only use sibling props via proxy or use `this`
- Otherwise use `derive`
?

**Reply by dai-shi** [reply]
> Now, I think I understand your recommendation on `this`:
> 
> * Use `this`
> 
> * Otherwise use `derive`

Yeah, that sounds about right.

> Still, will it be better to recommend:
> 
> * Only use sibling props via proxy or use `this`

Can you elaborate on it? Can you share some very small examples that is recommended and not recommended?

**Reply by alexander-entin** [reply]
````javascript
user = proxy({
    name: 'John',
    get greetingEn() { return 'Hello ' + user.name }, // can reference sibling props via the proxy object - OK
    get greetingFr() { return 'Bonjour ' + this.name }, // via `this` - OK
})

state = proxy({
    user: { // could be nested
        name: 'John',
        get greetingEn() { return 'Hello ' + state.user.name }, // can reference sibling props via the proxy object - OK
        get greetingFr() { return 'Bonjour ' + this.name }, // via `this` - OK
    },
})

state = proxy({
    user: {
        name: 'John',
    },
    greetings: {
        get greetingEn() { return 'Hello ' + state.user.name }, // WRONG, can reference sibling props only. Use `derive` instead
        get greetingFr() { return 'Bonjour ' + this.user.name }, // WRONG, `this` points to `state.greetings`. Use `derive` instead
    },
})

user = proxy({ 
    name: 'John',
})
greetings = proxy({
    get greetingEn() { return 'Hello ' + user.name }, // WRONG, can reference sibling props only. Use `derive` instead
    get greetingFr() { return 'Bonjour ' + this.name }, // WRONG, `this` points to `greetings`. Use `derive` instead
})
````

**Reply by dai-shi** [reply]
```js
const user = proxy({
    name: 'John',
    get greetingEn() { return 'Hello ' + user.name }, // this is NG, because snapshot refers to mutable variable
    get greetingFr() { return 'Bonjour ' + this.name }, // this is OK, because snapshot is a snapshot (frozen)
})

const userSnap = snapshot(user)

user.name += '!'

userSnap.greetingEn // ---> `Hello John!` which is not expected
userSnap.greetingFr // ---> `Bonjour John` which is correct
```

**Reply by alexander-entin** [reply]
except that `userSnap.greetingEn` returns `Hello John` (w/o bang)
<img width="701" alt="1" src="https://github.com/pmndrs/valtio/assets/10639790/2efbac6e-c1ef-4e7f-b6be-61e8773ebb22">

````
    "node_modules/valtio": {
      "version": "1.10.4",
      "resolved": "https://registry.npmjs.org/valtio/-/valtio-1.10.4.tgz",
      "integrity": "sha512-gqGWh0DjtDMAy8Jaui8ufFoxlQB1k1NiA/QHrpKoTUk9EeY331WKeYhvtGn1u703RcefrDCez7PT+qeCu9lWEw==",
      "dependencies": {
        "proxy-compare": "2.5.0",
        "use-sync-external-store": "1.2.0"
      },
````

Also, official guide provides a very similar example
````javascript
const state = proxy({
  count: 1,
  get doubled() {
    // We recommend `state.count` for clarity to beginners, but `this.count` works too
    return state.count * 2
  },
})
console.log(state.doubled) // 2

// Getter calls on the snapshot work as expected
const snap = snapshot(state)
console.log(snap.doubled) // 2

// When count changes in the state proxy
state.count = 10
// Then snapshot's computed property does not change
console.log(snap.doubled) // 2
````
And an explanation:
> when you make a snapshot, calls to snap.doubled are effectively cached, because the value of an object getter is copied during the snapshot process

**Reply by dai-shi** [reply]
> except that userSnap.greetingEn returned `Hello John` (w/o bang)

Oops!!! You are right, that was my bad. It's a getter function. I confused with a normal function.
Yes, the snapshot object is created with the getter result.
Sorry about my wrong comment.

**Reply by dai-shi** [reply]
> Also, official guide provides a very similar example

Oh, yes... It was tricky for me.
Now, I think you have good understanding, what would be your suggestion to improve docs?

**Reply by alexander-entin** [reply]
I suggest adding the following to the end of "Simple object getter" section of https://valtio.pmnd.rs/docs/guides/computed-properties
(not exact wording, but the idea):

> Please note, that in the current implementation a computed property should only reference _sibling_ properties, otherwise you'll encounter weird bugs. For example:
(and examples from https://github.com/pmndrs/valtio/discussions/731#discussioncomment-5955256)

**Reply by dai-shi** [reply]
Looks good. Would you like to open a PR? Or, want someone to work on it instead?

**Reply by alexander-entin** [reply]
Maybe it's time to learn how to do it

**Reply by dai-shi** [reply]
Let us know if you need any help. You can take a look at previous PRs.

**Comment by ShookLyngs** [comment]
What would happen if we do this:
```
const a = proxy({
  type: 'a',
  address: '0x00',
});
const b = proxy({
  type: 'b',
  address: '0x01',
});
const wallet = proxy({
  instance: a,
  get type() {
    return wallet.instance.type;
  },
  get address() {
    return wallet.instance.address;
  }
});

wallet.instance = b;
```

**Reply by dai-shi** [reply]
it should work the same as without `proxy()`.
